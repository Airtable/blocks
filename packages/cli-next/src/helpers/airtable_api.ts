import fetch from 'node-fetch';
import {spawnUnexpectedError, spawnUserError} from './error_utils';
import {FetchApi, Response, FetchInit} from './fetch_api';
import {S3Api, S3SignedUploadInfo} from './s3_api';

/**
 * Enumeration of some error types server may return.
 */
/* eslint-disable airtable/enum-style */
enum ErrorResponseCode {
    RESOURCE_NOT_FOUND = 'RESOURCE_NOT_FOUND',
    UNSUPPORTED_BLOCKS_CLI_VERSION = 'UNSUPPORTED_BLOCKS_CLI_VERSION',
}
/* eslint-enable airtable/enum-style */

export enum AirtableApiErrorName {
    AIRTABLE_API_ERROR_STATUS_AND_MESSAGES = 'airtableApiErrorStatusAndMessages',
    AIRTABLE_API_KEY_MALFORMED = 'airtableApiKeyMalformed',
    AIRTABLE_API_KEY_NAME_INVALID = 'airtableApiKeyNameInvalid',
    AIRTABLE_API_MULTIPLE_ERRORS = 'airtableApiMultipleErrors',
    AIRTABLE_API_WITH_INVALID_API_KEY = 'airtableApiWithInvalidApiKey',
    AIRTABLE_API_BLOCK_NOT_FOUND = 'airtableApiBlockNotFound',
    AIRTABLE_API_UNSUPPORTED_BLOCKS_CLI_VERSION = 'airtableApiUnsupportedBlocksCliVersion',
    AIRTABLE_API_UNEXPECTED_ERROR = 'airtableApiUnexpectedError',
}

export interface AirtableApiErrorStatusAndMessages {
    type: AirtableApiErrorName.AIRTABLE_API_ERROR_STATUS_AND_MESSAGES;
    status: number;
    errors: Exclude<ErrorResponseJson['errors'], undefined>;
}

export interface AirtableApiErrorKeyMalformed {
    type: AirtableApiErrorName.AIRTABLE_API_KEY_MALFORMED;
}

export interface AirtableApiErrorKeyNameInvalid {
    type: AirtableApiErrorName.AIRTABLE_API_KEY_NAME_INVALID;
    name: string;
}

export interface AirtableApiErrorMultiple {
    type: AirtableApiErrorName.AIRTABLE_API_MULTIPLE_ERRORS;
    errors: AirtableApiErrorInfo[];
}

export interface AirtableApiErrorInvalidApiKey {
    type: AirtableApiErrorName.AIRTABLE_API_WITH_INVALID_API_KEY;
}

export interface AirtableApiErrorBaseNotFound {
    type: AirtableApiErrorName.AIRTABLE_API_BLOCK_NOT_FOUND;
}

export interface AirtableApiErrorUnsupportedBlocksCliVersion {
    type: AirtableApiErrorName.AIRTABLE_API_UNSUPPORTED_BLOCKS_CLI_VERSION;
    serverMessage: string;
}

export interface AirtableApiErrorUnexpected {
    type: AirtableApiErrorName.AIRTABLE_API_UNEXPECTED_ERROR;
    serverMessage: string;
}

export type AirtableApiErrorInfo =
    | AirtableApiErrorStatusAndMessages
    | AirtableApiErrorKeyMalformed
    | AirtableApiErrorKeyNameInvalid
    | AirtableApiErrorMultiple
    | AirtableApiErrorInvalidApiKey
    | AirtableApiErrorBaseNotFound
    | AirtableApiErrorUnsupportedBlocksCliVersion
    | AirtableApiErrorUnexpected;

interface ErrorResponseJson {
    error?: {
        code?: ErrorResponseCode | string;
        message: string;
    };
    errors?: {
        code?: ErrorResponseCode | string;
        message: string;
    }[];
}

interface ErrorResponseError {
    status: number;
    error: {
        message: string;
        code?: string;
    };
}

export interface AirtableApiOptions {
    blockId: string;
    apiKey: string;
    userAgent: string;
    apiBaseUrl: string;
}

export interface CreateBuildOptions {
    s3: S3Api;
    frontendBundle: Buffer;
    backendBundle: Buffer | null;
}

export interface CreateReleaseOptions {
    buildId: string;
    deployId?: string;
    developerComment?: string;
}

export interface CodeUploadOptions {
    codeUploadId: string;
    status: 'uploaded' | 'failed';
}

export interface CodeUploadResponse {
    codeUploadId: string;
    presignedUploadUrl: string;
}

export interface FinalizeCodeUploadResponse {
    message: string;
}

export interface UploadSubmissionOptions {
    archiveBuffer: Buffer;
}

export interface CreateBuildResponseJson {
    buildId: string;
    frontendBundleUploadUrl: string;
    backendDeploymentPackageUploadUrl: string | null;
    frontendBundleS3UploadInfo: S3SignedUploadInfo;
    backendDeploymentPackageS3UploadInfo: S3SignedUploadInfo | null;
}

function createAirtableApiError({
    status,
    error: {message, code},
}: ErrorResponseError): AirtableApiErrorInfo {
    if (status === 401) {
        return {
            type: AirtableApiErrorName.AIRTABLE_API_WITH_INVALID_API_KEY,
        };
    } else if (status === 422 && code === ErrorResponseCode.RESOURCE_NOT_FOUND) {
        return {
            type: AirtableApiErrorName.AIRTABLE_API_BLOCK_NOT_FOUND,
        };
    } else {
        return {
            type: AirtableApiErrorName.AIRTABLE_API_UNEXPECTED_ERROR,
            serverMessage: message,
        };
    }
}

export abstract class AirtableApi extends FetchApi {
    protected blockId: string;
    private apiKey: string;
    private userAgent: string;
    private apiBaseUrl: string;

    constructor({blockId, apiKey, userAgent, apiBaseUrl}: AirtableApiOptions) {
        super();
        this.blockId = blockId;
        this.apiKey = apiKey;
        this.userAgent = userAgent;
        this.apiBaseUrl = apiBaseUrl;
    }

    airtableFetchInit({url, ...init}: FetchInit): FetchInit {
        return {
            url: new URL(url, this.apiBaseUrl).href,
            ...init,
            method: 'post',
            headers: {
                Authorization: `Bearer ${this.apiKey}`,
                'User-Agent': this.userAgent,
                'Content-Type': 'application/json',
            },
        };
    }

    abstract async createBuildAsync(options: CreateBuildOptions): Promise<CreateBuildResponseJson>;

    abstract async createReleaseAsync(options: CreateReleaseOptions): Promise<void>;

    async uploadSubmissionAsync({archiveBuffer}: UploadSubmissionOptions): Promise<string> {
        const {presignedUploadUrl, codeUploadId} = await this._blockCreateCodeUploadAsync();

        const response = await fetch(presignedUploadUrl, {method: 'put', body: archiveBuffer});

        const didUpload = response.status === 200;
        const status = didUpload ? 'uploaded' : 'failed';
        const {message} = await this._blockFinalizeCodeUploadAsync({
            codeUploadId,
            status,
        });

        if (!didUpload) {
            throw spawnUnexpectedError(message);
        }

        return message;
    }

    protected abstract async _blockCreateCodeUploadAsync(): Promise<CodeUploadResponse>;

    protected abstract async _blockFinalizeCodeUploadAsync(
        options: CodeUploadOptions,
    ): Promise<FinalizeCodeUploadResponse>;

    protected async _invariantOkResponseAsync(init: FetchInit, response: Response) {
        const {status} = response;
        if (status === 200) {
            return;
        }

        const {error, errors}: ErrorResponseJson = await response.json();
        if (Array.isArray(errors) && errors.length > 1) {
            if (errors.some(errObj => errObj && typeof errObj.message === 'string')) {
                throw spawnUserError<AirtableApiErrorInfo>({
                    type: AirtableApiErrorName.AIRTABLE_API_ERROR_STATUS_AND_MESSAGES,
                    status,
                    errors,
                });
            }
            const errorMessages = errors.map(err => createAirtableApiError({status, error: err}));
            throw spawnUserError<AirtableApiErrorInfo>({
                type: AirtableApiErrorName.AIRTABLE_API_MULTIPLE_ERRORS,
                errors: errorMessages,
            });
        } else {
            const errorToThrow = Array.isArray(errors) && errors.length === 1 ? errors[0] : error;
            if (
                errorToThrow &&
                errorToThrow.code === ErrorResponseCode.UNSUPPORTED_BLOCKS_CLI_VERSION &&
                typeof errorToThrow.message === 'string'
            ) {
                throw spawnUserError<AirtableApiErrorInfo>({
                    type: AirtableApiErrorName.AIRTABLE_API_UNSUPPORTED_BLOCKS_CLI_VERSION,
                    serverMessage: errorToThrow.message,
                });
            } else if (errorToThrow && typeof errorToThrow.message === 'string') {
                throw spawnUserError<AirtableApiErrorInfo>(
                    createAirtableApiError({status, error: errorToThrow}),
                );
            } else {
                throw spawnUnexpectedError(
                    'Request to Airtable failed with status code %s',
                    status,
                );
            }
        }
    }
}
