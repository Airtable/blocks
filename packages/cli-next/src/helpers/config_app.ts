import {spawnUserError, UserError} from './error_utils';
import {Result} from './result';

export enum AppConfigErrorName {
    APP_CONFIG_IS_NOT_VALID = 'appConfigIsNotValid',
}

export interface AppConfigErrorInvalid {
    type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID;
    file?: string;
    message: string;
}

export type AppConfigErrorInfo = AppConfigErrorInvalid;

export interface AppBundlerConfig {
    /** Resolvable module to run to bundle the block. */
    readonly module?: string;

    /** Options passed to the bundler by the CLI. */
    readonly options?: null;
}

export interface AppConfig {
    readonly version?: '1.0';

    /** Frontend entry file for application. */
    readonly frontendEntry: string;

    /** Swappable bundler extension and options. */
    readonly bundler?: AppBundlerConfig;
}

export function validateAppConfig(
    value: unknown,
): Result<AppConfig, UserError<AppConfigErrorInfo>> {
    if (typeof value !== 'object' || value === null) {
        return {
            err: spawnUserError({
                type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID,
                message: 'should be a non-null object',
            }),
        };
    } else if (Array.isArray(value)) {
        return {
            err: spawnUserError({
                type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID,
                message: 'should be a non-array object',
            }),
        };
    }

    const config = value as AppConfig;
    if (!['1.0', undefined].includes(config.version)) {
        return {
            err: spawnUserError({
                type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID,
                message: 'version should be "1.0" or undefined',
            }),
        };
    }

    if (typeof config.frontendEntry !== 'string') {
        return {
            err: spawnUserError({
                type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID,
                message: 'frontendEntry should be a string',
            }),
        };
    }

    if ('bundler' in config) {
        if (typeof config.bundler === 'object' && config.bundler !== null) {
            if ('module' in config.bundler && typeof config.bundler.module !== 'string') {
                return {
                    err: spawnUserError({
                        type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID,
                        message: 'bundler.module should be a string or not set',
                    }),
                };
            }
            if ('options' in config.bundler && typeof config.bundler.options !== 'object') {
                return {
                    err: spawnUserError({
                        type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID,
                        message: 'bundler.options should be an object',
                    }),
                };
            }
        } else {
            return {
                err: spawnUserError({
                    type: AppConfigErrorName.APP_CONFIG_IS_NOT_VALID,
                    message: 'bundler should be an object',
                }),
            };
        }
    }

    return {value: config};
}

export function appConfigGetFrontendEntry(appConfig: AppConfig): string {
    return appConfig.frontendEntry;
}
