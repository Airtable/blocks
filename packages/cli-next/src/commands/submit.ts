import {flags as commandFlags} from '@oclif/command';
import _debug from 'debug';
import archiver from 'archiver';

import cli from '../helpers/cli_ux';

import AirtableCommand from '../helpers/airtable_command';
import {
    AIRTABLE_API_URL,
    APP_RELEASE_DIR,
    APP_ROOT_TEMPORARY_DIR,
    SUBMIT_ARCHIVE_NAME,
    V2_BLOCKS_BASE_ID,
} from '../settings';

import {AppBundlerContext, resolveBundlerModuleAsync} from '../manager/bundler';
import {createSubmitTaskAsync, SubmitTaskProducer} from '../manager/submit';
import {SubmitTaskConsumerAdapter} from '../manager/submit_adapter';

import {
    findAppConfigAsync,
    findAppDirectoryAsync,
    findRemoteConfigPathByNameAsync,
    readAppConfigAsync,
    readRemoteConfigAsync,
    validateRemoteName,
} from '../helpers/system_config';
import {renderEntryPointAsync} from '../helpers/render_entry_point_async';
import {mkdirpAsync, rmdirAsync, unlinkIfExistsAsync} from '../helpers/system_extra';
import {readApiKeyAsync} from '../helpers/system_api_key';
import {createUserAgentAsync} from '../helpers/user_agent';
import {Deferred} from '../helpers/deferred';
import {unwrapResultFunctor} from '../helpers/result';
import {System} from '../helpers/system';
import {SubmitCommandErrorName, SubmitCommandMessageName} from '../helpers/submit_messages';
import {spawnUserError, spawnUnexpectedError} from '../helpers/error_utils';
import {AirtableLegacyBlockApi} from '../helpers/airtable_legacy_block_api';
import {AirtableBlockV2Api} from '../helpers/airtable_block_v2_api';
import {AirtableApi} from '../helpers/airtable_api';
import {resolveBuiltinModuleAsync} from '../helpers/task';

const debug = _debug('block-cli:command:submit');

class SubmitProducer implements SubmitTaskProducer {
    readyDefer = new Deferred<void>();

    async readyAsync() {
        this.readyDefer.resolve();
    }
}

export default class Submit extends AirtableCommand {
    private _task?: SubmitTaskConsumerAdapter;
    private _teardownAction?: (() => void) | null;
    /**
     * A file system path describing the location where a temporary directory
     * should be created to store the generated file(s).
     */
    private _appTemporaryPath?: string;

    static description = 'Submit extension for review for listing in the the Airtable Marketplace';

    static examples = [
        `$ block submit
`,
    ];

    static flags = {
        help: commandFlags.help({char: 'h'}),
        remote: commandFlags.string({
            description: 'Configure which remote to use',
            parse: unwrapResultFunctor(validateRemoteName),
        }),
    };

    async runAsync() {
        const {flags} = this.parse(Submit);

        const sys = this.system;
        const workingdir = sys.process.cwd();

        const appRootPath = await findAppDirectoryAsync(sys, workingdir);
        const appConfigPath = await findAppConfigAsync(sys, appRootPath);
        const appConfigResult = await readAppConfigAsync(sys, appConfigPath);
        if (appConfigResult.err) {
            this.error(appConfigResult.err);
        }
        const appConfig = appConfigResult.value;
        debug('loaded extension config at %s', sys.path.relative(workingdir, appConfigPath));

        const remoteConfigPath = await findRemoteConfigPathByNameAsync(
            sys,
            workingdir,
            flags.remote,
        );
        const remoteConfigResult = await readRemoteConfigAsync(sys, remoteConfigPath);
        if (remoteConfigResult.err) {
            this.error(remoteConfigResult.err);
        }
        const remoteConfig = remoteConfigResult.value;
        const {baseId, blockId} = remoteConfig;
        debug('loaded remote config at %s', sys.path.relative(workingdir, remoteConfigPath));

        const apiKeyResult = await readApiKeyAsync(sys, remoteConfig.apiKeyName);
        if (apiKeyResult.err) {
            this.error(apiKeyResult.err);
        }
        const apiKey = apiKeyResult.value;
        debug('loaded apiKey %s', remoteConfig.apiKeyName ?? 'default');

        const userAgent = await createUserAgentAsync(sys);
        debug('connecting to Airtable with user agent: %s', userAgent);

        const bundlerContext: AppBundlerContext = {
            module: appConfig.bundler?.module,
            workingdir: appRootPath,
        };
        const bundlerModule = await resolveBundlerModuleAsync(sys, bundlerContext);
        const defaultBundlerModule = await resolveBuiltinModuleAsync(
            sys,
            __dirname,
            '..',
            'bundler',
        );

        const producer = new SubmitProducer();
        const task = await createSubmitTaskAsync(sys, bundlerContext, producer);
        this._task = task;
        debug('initialized task');

        this._appTemporaryPath = sys.path.join(appRootPath, APP_ROOT_TEMPORARY_DIR);
        await mkdirpAsync(sys, this._appTemporaryPath);
        debug('created temporary directory');

        const appBuildPath = sys.path.join(this._appTemporaryPath, APP_RELEASE_DIR);
        await mkdirpAsync(sys, appBuildPath);
        debug('created build directory');

        cli.action.start('Finding dependencies');
        this._teardownAction = () => {
            this._teardownAction = null;
            cli.action.stop('Incomplete');
        };

        const entryPointPath = sys.path.join(this._appTemporaryPath, 'index.js');
        const userEntryPoint = sys.path.join(appRootPath, appConfig.frontendEntry);
        const entryPoint = await renderEntryPointAsync(sys, {
            mode: 'production',
            destination: entryPointPath,
            userEntryPoint,
        });
        await sys.fs.writeFileAsync(entryPointPath, Buffer.from(entryPoint));
        debug('wrote generated entry file at %s', sys.path.relative(workingdir, entryPointPath));

        await producer.readyDefer.promise;

        debug('finding dependencies');
        const dependencies = await task.findDependenciesAsync({
            mode: 'production',
            context: appRootPath,
            entry: entryPointPath,
        });
        debug('found %d dependencies', dependencies.files.length);

        cli.action.stop();
        cli.action.start('Packaging');

        const projectItems = await projectFilesAsync(sys, appRootPath);
        debug('project has %d descendant files', projectItems.length);

        const customBundlerModules = bundlerModule === defaultBundlerModule ? [] : [bundlerModule];
        debug('using custom bundler %s', bundlerModule);

        const uniqueItems = [
            ...projectItems,
            ...dependencies.files,
            ...customBundlerModules,
        ].reduce((carry, file) => {
            if (!carry.includes(file)) {
                carry.push(file);
            }
            return carry;
        }, [] as string[]);

        const filesOnly = (
            await Promise.all(
                uniqueItems.map(file =>
                    sys.fs
                        .readdirAsync(file)
                        .then(() => null)
                        .catch(() => file),
                ),
            )
        ).filter(Boolean) as string[];

        const commonRootPath = filesOnly
            .map(file => sys.path.dirname(file))
            .reduce((carry, dir) => {
                while (!dir.startsWith(carry)) {
                    const lastCarry = carry;
                    carry = sys.path.dirname(carry);
                    if (lastCarry === carry) {
                        throw spawnUserError({
                            type: SubmitCommandErrorName.SUBMIT_COMMAND_WINDOWS_MULTIPLE_DISKS,
                        });
                    }
                }
                return carry;
            });
        const commonRootRelativePath = sys.path.relative(workingdir, commonRootPath);

        const outputPath = sys.path.join(appBuildPath, SUBMIT_ARCHIVE_NAME);
        const outputRelativePath = sys.path.relative(workingdir, outputPath);
        await unlinkIfExistsAsync(sys, outputPath);
        const output = sys.fs.createWriteStream(outputPath);
        debug('writing zip to %s', outputRelativePath);

        const zip = archiver('zip');
        zip.pipe(output);
        zip.append(Buffer.from(sys.path.relative(commonRootPath, appRootPath)), {
            name: 'AIRTABLE_BLOCK_ROOT_PATH',
        });
        const zipFileTable = ['AIRTABLE_BLOCK_ROOT_PATH'];
        for (const file of filesOnly) {
            const name = sys.path.relative(commonRootPath, file);
            zipFileTable.push(name);
            zip.append(sys.fs.createReadStream(file), {
                name,
            });
        }
        const fileStreamClosedPromise = new Promise(resolve => {
            output.on('close', resolve);
        });
        await zip.finalize();
        await fileStreamClosedPromise;
        await sys.fs.writeFileAsync(
            sys.path.join(appBuildPath, 'block_archive.files.txt'),
            Buffer.from(`${zipFileTable.join('\n')}\n`),
        );

        const archiveBuffer = await sys.fs.readFileAsync(
            sys.path.join(appBuildPath, SUBMIT_ARCHIVE_NAME),
        );
        debug(
            'zipped up %d bytes from %d files under the common directory %s',
            archiveBuffer.byteLength,
            filesOnly.length,
            commonRootRelativePath,
        );

        cli.action.stop();

        const check = await cli.prompt(
            this.messages.renderMessage({
                type: SubmitCommandMessageName.SUBMIT_COMMAND_PACKAGED_CONTINUE_PROMPT,
            }),
            {
                default: 'N',
            },
        );
        if (check === 'N') {
            this.logMessage({type: SubmitCommandMessageName.SUBMIT_COMMAND_STOP_AFTER_PACKAGING});
            return;
        }

        cli.action.start('Submitting');

        let airtableApi: AirtableApi;
        if (baseId === V2_BLOCKS_BASE_ID) {
            airtableApi = new AirtableBlockV2Api({
                blockId,
                apiKey,
                userAgent,
                apiBaseUrl: remoteConfig.server ?? AIRTABLE_API_URL,
            });
        } else {
            airtableApi = new AirtableLegacyBlockApi({
                baseId,
                blockId,
                apiKey,
                userAgent,
                apiBaseUrl: remoteConfig.server ?? AIRTABLE_API_URL,
            });
        }
        const submitResponseMessage = await airtableApi.uploadSubmissionAsync({
            archiveBuffer,
        });

        cli.action.stop();
        this._teardownAction = null;

        this.log(submitResponseMessage);
    }

    async finallyAsync() {
        await Promise.all([
            this._appTemporaryPath ? rmdirAsync(this.system, this._appTemporaryPath) : null,
            this._task ? this._task.teardownAsync() : null,
            this._teardownAction ? this._teardownAction() : null,
        ]);
    }
}

const excludeProjectItems = ['node_modules', 'build', '.tmp', '.git'];
async function projectFilesAsync(sys: System, dir: string): Promise<string[]> {
    if (excludeProjectItems.includes(sys.path.parse(dir).base)) {
        return [];
    }

    const stat = await sys.fs.statAsync(dir);
    if (stat.isFile()) {
        return [dir];
    } else if (stat.isDirectory()) {
        const entries = await sys.fs.readdirAsync(dir);
        const nested = await Promise.all(
            entries.map(entry => projectFilesAsync(sys, sys.path.join(dir, entry))),
        );
        return ([] as string[]).concat(...nested);
    } else {
        throw spawnUnexpectedError('Unexpected object %s', dir);
    }
}
