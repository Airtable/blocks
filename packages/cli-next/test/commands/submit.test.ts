import {expect, test} from '../mocks/test';

import * as submitModule from '../../src/manager/submit';
import * as airtableLegacyBlockApiModule from '../../src/helpers/airtable_legacy_block_api';
import * as airtableBlockV2ApiModule from '../../src/helpers/airtable_block_v2_api';
import * as taskModule from '../../src/helpers/task';
import * as userAgentModule from '../../src/helpers/user_agent';

import {AppConfigErrorName} from '../../src/helpers/config_app';
import {RemoteConfigErrorName} from '../../src/helpers/config_remote';
import {SubmitCommandMessageName} from '../../src/helpers/submit_messages';
import {SubmitTaskConsumerAdapter} from '../../src/manager/submit_adapter';
import {System} from '../../src/helpers/system';
import {SystemApiKeyErrorName} from '../../src/helpers/system_api_key';

import {mapFancyTestAsyncPlugin} from '../mocks/FancyTestAsync';
import {spawnUnexpectedError} from '../../src/helpers/error_utils';
import {UploadSubmissionOptions} from '../../src/helpers/airtable_api';

type SubmitTaskProducer = submitModule.SubmitTaskProducer;

const {
    stubResolveBuiltinModuleAsync,
    stubCreateSubmitTaskAsync,
    airtableLegacyBlockApiStub,
    airtableBlockV2ApiStub,
} = createStubs();

describe('submit', () => {
    const testSubmitCommand = test
        .register('wroteFileMatchingSnapshot', wroteFileMatchingSnapshot)
        .timeout(10000)
        .stdout()
        .stderr()
        .enableDebug('block-cli*:submit')
        .withJSON({
            '/home/.config/.airtableblocksrc.json': {airtableApiKey: {default: 'key1234'}},
            '/home/projects/my-app/.block/remote.json': {baseId: 'app1234', blockId: 'blk5678'},
            '/home/projects/my-app/.block/newremote.json': {baseId: 'app5678', blockId: 'blk1234'},
            '/home/projects/my-app/block.json': {frontendEntry: 'frontend/index.js'},
            '/home/projects/my-app/package.json': {},
            '/home/projects/shared/package.json': {},
        })
        .withFiles({
            '/node/env-12.20.1/lib/node_modules/@airtable/blocks-cli/lib/bundler.js': Buffer.from(
                '',
            ),
            '/home/projects/my-app/frontend/index.js': Buffer.from(''),
            '/home/projects/shared/lib/index.js': Buffer.from(''),
        })
        .stubDirectoryRemoval()
        .stub(taskModule, 'resolveBuiltinModuleAsync', stubResolveBuiltinModuleAsync())
        .stub(
            submitModule,
            'createSubmitTaskAsync',
            stubCreateSubmitTaskAsync(['/home/projects/my-app/frontend/index.js']),
        )
        .stub(airtableLegacyBlockApiModule, 'AirtableLegacyBlockApi', airtableLegacyBlockApiStub())
        .stub(
            airtableBlockV2ApiModule,
            'AirtableBlockV2Api',
            airtableBlockV2ApiStub({
                constructorOptions() {
                    throw spawnUnexpectedError('Must use AirtableLegacyBlockApi');
                },
            }),
        )
        .stub(userAgentModule, 'createUserAgentAsync', () => 'airtable-cli-user-agent/1.0.0');

    const testSubmitCommandAndContinue = testSubmitCommand.answer(
        SubmitCommandMessageName.SUBMIT_COMMAND_PACKAGED_CONTINUE_PROMPT,
        {stdin: 'y'},
    );

    testSubmitCommandAndContinue
        .command(['submit'])
        .wroteFileMatchingSnapshot('/home/projects/my-app/.tmp/dist/block_archive.files.txt')
        .it('completes', ({stderr}) => {
            expect(stderr).to.contain('Submitting');
        });

    testSubmitCommand
        .answer(SubmitCommandMessageName.SUBMIT_COMMAND_PACKAGED_CONTINUE_PROMPT, {stdin: ''})
        .command(['submit'])
        .it('stops on request', ({stdout}) => {
            expect(stdout).to.contain(SubmitCommandMessageName.SUBMIT_COMMAND_STOP_AFTER_PACKAGING);
        });

    testSubmitCommandAndContinue
        .stub(
            submitModule,
            'createSubmitTaskAsync',
            stubCreateSubmitTaskAsync([
                '/home/projects/my-app/frontend/index.js',
                '/home/projects/shared/lib/index.js',
            ]),
        )
        .command(['submit'])
        .wroteFileMatchingSnapshot('/home/projects/my-app/.tmp/dist/block_archive.files.txt')
        .it('completes with sibling shared source');

    testSubmitCommandAndContinue
        .withFiles({
            '/home/projects/my-app/my-bundler.js': Buffer.from(''),
        })
        .withJSON({
            '/home/projects/my-app/block.json': {
                frontendEntry: 'frontend/index.js',
                bundler: {module: './my-bundler.js'},
            },
        })
        .command(['submit'])
        .wroteFileMatchingSnapshot('/home/projects/my-app/.tmp/dist/block_archive.files.txt')
        .it('completes with custom bundler');

    testSubmitCommandAndContinue
        .withFiles({
            '/home/projects/shared/my-bundler.js': Buffer.from(''),
        })
        .withJSON({
            '/home/projects/my-app/block.json': {
                frontendEntry: 'frontend/index.js',
                bundler: {module: '../shared/my-bundler.js'},
            },
        })
        .command(['submit'])
        .wroteFileMatchingSnapshot('/home/projects/my-app/.tmp/dist/block_archive.files.txt')
        .it('completes with custom bundler in shared dir');

    testSubmitCommandAndContinue
        .withJSON({
            '/home/.config/.airtableblocksrc.json': {
                airtableApiKey: {default: 'key1234', another: 'key5678'},
            },
            '/home/projects/my-app/.block/newremote.remote.json': {
                baseId: 'app5678',
                blockId: 'blk1234',
                server: 't.t.com',
                apiKeyName: 'another',
            },
        })
        .stub(
            airtableLegacyBlockApiStub,
            'AirtableLegacyBlockApi',
            airtableLegacyBlockApiStub({
                constructorOptions(
                    options: airtableLegacyBlockApiModule.AirtableLegacyBlockApiBaseOptions,
                ) {
                    expect(options.baseId).to.equal('app5678');
                    expect(options.apiBaseUrl).to.equal('t.t.com');
                    expect(options.apiKey).to.equal('key5678');
                },
            }),
        )
        .stub(
            airtableBlockV2ApiModule,
            'AirtableBlockV2Api',
            airtableBlockV2ApiStub({
                constructorOptions() {
                    throw spawnUnexpectedError('Must use AirtableLegacyBlockApi');
                },
            }),
        )
        .command(['submit', '--remote', 'newremote'])
        .it('submits to newremote remote');

    testSubmitCommand
        .withJSON({
            '/home/projects/my-app/block.json': {},
        })
        .command('submit')
        .catch(new RegExp(AppConfigErrorName.APP_CONFIG_IS_NOT_VALID))
        .it('validates block.json');

    testSubmitCommand
        .withJSON({
            '/home/projects/my-app/.block/remote.json': {},
        })
        .command('submit')
        .catch(new RegExp(RemoteConfigErrorName.REMOTE_CONFIG_IS_NOT_VALID))
        .it('validates remote.json');

    testSubmitCommand
        .withJSON({
            '/home/.config/.airtableblocksrc.json': {},
        })
        .command('submit')
        .catch(new RegExp(SystemApiKeyErrorName.SYSTEM_API_KEY_NOT_FOUND))
        .it('ensures apiKey is in user config');
});

function createStubs() {
    function _stubResolveBuiltinModuleAsync(): any {
        return async function(
            sys: System,
            workingDir: string,
            ...modulePath: string[]
        ): Promise<string> {
            return '/node/env-12.20.1/lib/node_modules/@airtable/blocks-cli/lib/bundler.js';
        };
    }

    function _stubCreateSubmitTaskAsync(files: string[] = []): any {
        return function(sys: System, context: any, producer: SubmitTaskProducer) {
            (async () => {
                await Promise.resolve();
                await producer.readyAsync();
            })();

            return {
                async findDependenciesAsync() {
                    return {files};
                },
                async teardownAsync() {},
            } as SubmitTaskConsumerAdapter;
        };
    }

    type AirtableLegacyBlockApiStubMethods = {
        constructorOptions(
            options: airtableLegacyBlockApiModule.AirtableLegacyBlockApiBaseOptions,
        ): void;
    } & Pick<airtableLegacyBlockApiModule.AirtableLegacyBlockApi, 'uploadSubmissionAsync'>;

    function _airtableLegacyBlockApiStub(
        methods: Partial<AirtableLegacyBlockApiStubMethods> = {},
    ): any {
        return class AirtableLegacyBlockApiStub {
            private options: airtableLegacyBlockApiModule.AirtableLegacyBlockApiBaseOptions;

            constructor(options: airtableLegacyBlockApiModule.AirtableLegacyBlockApiBaseOptions) {
                if (methods.constructorOptions) {
                    methods.constructorOptions(options);
                }
                this.options = options;
            }

            async uploadSubmissionAsync(options: UploadSubmissionOptions) {
                if (methods.uploadSubmissionAsync) {
                    return await methods.uploadSubmissionAsync(options);
                }
                return 'message';
            }
        };
    }

    type AirtableBlockV2ApiStubMethods = {
        constructorOptions(options: airtableBlockV2ApiModule.AirtableBlockV2ApiBaseOptions): void;
    } & Pick<airtableBlockV2ApiModule.AirtableBlockV2Api, 'uploadSubmissionAsync'>;

    function _airtableBlockV2ApiStub(methods: Partial<AirtableBlockV2ApiStubMethods> = {}): any {
        return class AirtableBlockV2ApiStub {
            constructor(options: airtableBlockV2ApiModule.AirtableBlockV2ApiBaseOptions) {
                if (methods.constructorOptions) {
                    methods.constructorOptions(options);
                }
            }

            async uploadSubmissionAsync(options: UploadSubmissionOptions) {
                if (methods.uploadSubmissionAsync) {
                    return await methods.uploadSubmissionAsync(options);
                }
                return 'message';
            }
        };
    }

    return {
        stubResolveBuiltinModuleAsync: _stubResolveBuiltinModuleAsync,
        stubCreateSubmitTaskAsync: _stubCreateSubmitTaskAsync,
        airtableLegacyBlockApiStub: _airtableLegacyBlockApiStub,
        airtableBlockV2ApiStub: _airtableBlockV2ApiStub,
    };
}

function wroteFileMatchingSnapshot(filepath: string) {
    return mapFancyTestAsyncPlugin({
        async runAsync({system: sys}: {system: System}) {
            const buffer = await sys.fs.readFileAsync(filepath);
            const file = buffer.toString();
            expect(file).to.matchSnapshot();
        },
    });
}
