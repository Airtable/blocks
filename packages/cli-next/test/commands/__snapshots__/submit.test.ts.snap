// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submit completes 1`] = `
"AIRTABLE_BLOCK_ROOT_PATH
.block/newremote.json
.block/remote.json
block.json
frontend/index.js
package.json
"
`;

exports[`submit completes with custom bundler 1`] = `
"AIRTABLE_BLOCK_ROOT_PATH
.block/newremote.json
.block/remote.json
block.json
frontend/index.js
my-bundler.js
package.json
"
`;

exports[`submit completes with custom bundler in shared dir 1`] = `
"AIRTABLE_BLOCK_ROOT_PATH
my-app/.block/newremote.json
my-app/.block/remote.json
my-app/block.json
my-app/frontend/index.js
my-app/package.json
shared/my-bundler.js
"
`;

exports[`submit completes with sibling shared source 1`] = `
"AIRTABLE_BLOCK_ROOT_PATH
my-app/.block/newremote.json
my-app/.block/remote.json
my-app/block.json
my-app/frontend/index.js
my-app/package.json
shared/lib/index.js
"
`;
